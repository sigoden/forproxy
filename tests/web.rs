mod common;

use crate::common::{
    build_client, build_proxy_client, fetch_subscribe, mask_text, start_http_server, start_proxy,
    HELLO_WORLD,
};

use anyhow::Result;
use async_http_proxy::http_connect_tokio;
use futures_util::SinkExt;
use proxyfor::server::WEB_PREFIX;
use tokio::net::TcpStream;
use tokio_tungstenite::tungstenite::Message;

#[tokio::test]
async fn test_web_index() -> Result<()> {
    let (proxy_addr, stop_proxy) = start_proxy(true).await?;

    let client = build_client()?;

    let res = client
        .get(format!(
            "http://localhost:{}{}",
            proxy_addr.port(),
            WEB_PREFIX
        ))
        .send()
        .await?;

    let status = res.status();
    let text = res.text().await?;
    assert_eq!(status, 200);
    assert!(text.contains("<title>proxyfor</title>"));

    let _ = stop_proxy.send(());
    Ok(())
}

#[tokio::test]
async fn test_list_traffics() -> Result<()> {
    let (proxy_addr, stop_proxy) = start_proxy(true).await?;
    let (server_addr, stop_server) = start_http_server().await?;

    let proxy_client = build_proxy_client(&proxy_addr.to_string())?;

    let res = proxy_client
        .get(format!("http://localhost:{}/hello", server_addr.port()))
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    let res = proxy_client
        .get(format!(
            "http://localhost:{}/hello/gzip",
            server_addr.port()
        ))
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    let res = proxy_client
        .post(format!("http://localhost:{}/echo", server_addr.port()))
        .header("content-type", "text/plain")
        .body(HELLO_WORLD)
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    let mut output = vec![];

    let client = build_client()?;
    let res = client
        .get(format!(
            "http://localhost:{}{}/traffics",
            proxy_addr.port(),
            WEB_PREFIX
        ))
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    output.push("/traffics".into());
    output.push(res.text().await?);

    let client = build_client()?;
    let res = client
        .get(format!(
            "http://localhost:{}{}/traffics?markdown",
            proxy_addr.port(),
            WEB_PREFIX
        ))
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    output.push("/traffics?markdown".into());
    output.push(res.text().await?);

    let client = build_client()?;
    let res = client
        .get(format!(
            "http://localhost:{}{}/traffics?har",
            proxy_addr.port(),
            WEB_PREFIX
        ))
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    output.push("/traffics?har".into());
    output.push(res.text().await?);

    let client = build_client()?;
    let res = client
        .get(format!(
            "http://localhost:{}{}/traffics?curl",
            proxy_addr.port(),
            WEB_PREFIX
        ))
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    output.push("/traffics?curl".into());
    output.push(res.text().await?);

    let output = output.join("\n\n");
    let output = mask_text(&output);

    insta::assert_snapshot!(output);

    let _ = stop_server.send(());
    let _ = stop_proxy.send(());
    Ok(())
}

#[tokio::test]
async fn test_get_traffic() -> Result<()> {
    let (proxy_addr, stop_proxy) = start_proxy(true).await?;
    let (server_addr, stop_server) = start_http_server().await?;

    let proxy_client = build_proxy_client(&proxy_addr.to_string())?;

    let res = proxy_client
        .post(format!("http://localhost:{}/echo", server_addr.port()))
        .header("content-type", "text/plain")
        .body(HELLO_WORLD)
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    let mut output = vec![];

    let id = 1;
    let client = build_client()?;
    let res = client
        .get(format!(
            "http://localhost:{}{}/traffic/{}",
            proxy_addr.port(),
            WEB_PREFIX,
            id
        ))
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    output.push("/traffic/:id".into());
    output.push(res.text().await?);

    let client = build_client()?;
    let res = client
        .get(format!(
            "http://localhost:{}{}/traffic/{}?markdown",
            proxy_addr.port(),
            WEB_PREFIX,
            id
        ))
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    output.push("/traffic/:id?markdown".into());
    output.push(res.text().await?);

    let client = build_client()?;
    let res = client
        .get(format!(
            "http://localhost:{}{}/traffic/{}?har",
            proxy_addr.port(),
            WEB_PREFIX,
            id
        ))
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    output.push("/traffic/:id?har".into());
    output.push(res.text().await?);

    let client = build_client()?;
    let res = client
        .get(format!(
            "http://localhost:{}{}/traffic/{}?curl",
            proxy_addr.port(),
            WEB_PREFIX,
            id
        ))
        .send()
        .await?;

    assert_eq!(res.status(), 200);

    output.push("/traffic/:id?curl".into());
    output.push(res.text().await?);

    let output = output.join("\n\n");
    let output = mask_text(&output);

    insta::assert_snapshot!(output);

    let _ = stop_server.send(());
    let _ = stop_proxy.send(());
    Ok(())
}

#[tokio::test]
async fn test_subscribe_traffics() -> Result<()> {
    let (proxy_addr, stop_proxy) = start_proxy(true).await?;
    let (server_addr, stop_server) = start_http_server().await?;

    let proxy_client = build_proxy_client(&proxy_addr.to_string())?;

    for _ in 0..2 {
        let res = proxy_client
            .post(format!("http://localhost:{}/echo", server_addr.port()))
            .header("content-type", "text/plain")
            .body(HELLO_WORLD)
            .send()
            .await?;

        assert_eq!(res.status(), 200);
    }

    let output = fetch_subscribe(
        &format!(
            "http://localhost:{}{}/subscribe/traffics",
            proxy_addr.port(),
            WEB_PREFIX,
        ),
        2,
    )
    .await?;

    let output = mask_text(&output);
    insta::assert_snapshot!(output);

    let _ = stop_server.send(());
    let _ = stop_proxy.send(());
    Ok(())
}

#[tokio::test]
async fn test_subscribe_websocket() -> Result<()> {
    let (proxy_addr, stop_proxy) = start_proxy(true).await?;
    let (server_addr, stop_server) = start_http_server().await?;

    let mut stream = TcpStream::connect(proxy_addr).await?;
    http_connect_tokio(
        &mut stream,
        &server_addr.ip().to_string(),
        server_addr.port(),
    )
    .await?;

    let (mut ws, res) =
        tokio_tungstenite::client_async(format!("ws://{}", server_addr), stream).await?;

    assert_eq!(res.status(), 101);

    ws.send(Message::Text("hello".to_string())).await?;
    tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
    ws.send(Message::Text("hello".to_string())).await?;

    let output = fetch_subscribe(
        &format!(
            "http://localhost:{}{}/subscribe/websocket/1",
            proxy_addr.port(),
            WEB_PREFIX,
        ),
        4,
    )
    .await?;

    let output = mask_text(&output);
    insta::assert_snapshot!(output);

    let _ = stop_server.send(());
    let _ = stop_proxy.send(());

    Ok(())
}
