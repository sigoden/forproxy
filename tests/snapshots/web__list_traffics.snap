---
source: tests/web.rs
expression: output
---
/traffics

[
  {
    "id": 1,
    "method": "GET",
    "uri": "http://localhost:<PORT>/hello",
    "status": 200,
    "size": 13,
    "time": <TIME>,
    "mime": null
  },
  {
    "id": 2,
    "method": "GET",
    "uri": "http://localhost:<PORT>/hello/gzip",
    "status": 200,
    "size": 33,
    "time": <TIME>,
    "mime": null
  },
  {
    "id": 3,
    "method": "POST",
    "uri": "http://localhost:<PORT>/echo",
    "status": 200,
    "size": 13,
    "time": <TIME>,
    "mime": "text/plain"
  }
]

/traffics?markdown


# GET http://localhost:<PORT>/hello 200

REQUEST HEADERS
```
accept: */*
host: localhost:<PORT>
content-length: 0
```

RESPONSE HEADERS
```
content-length: 13
date: <DATETIME>
```

RESPONSE BODY
```
Hello, World!
```


# GET http://localhost:<PORT>/hello/gzip 200

REQUEST HEADERS
```
accept: */*
host: localhost:<PORT>
content-length: 0
```

RESPONSE HEADERS
```
content-encoding: gzip
transfer-encoding: chunked
date: <DATETIME>
```

RESPONSE BODY
```
Hello, World!
```


# POST http://localhost:<PORT>/echo 200

REQUEST HEADERS
```
content-type: text/plain
accept: */*
host: localhost:<PORT>
content-length: 13
```

REQUEST BODY
```plain
Hello, World!
```

RESPONSE HEADERS
```
content-type: text/plain
content-length: 13
date: <DATETIME>
```

RESPONSE BODY
```plain
Hello, World!
```

/traffics?har

{
  "log": {
    "version": "1.2",
    "creator": {
      "name": "proxyfor",
      "version": "0.2.0",
      "comment": ""
    },
    "pages": [],
    "entries": [
      {
        "startedDateTime": "<DATETIME>",
        "time": <TIME>,
        "request": {
          "method": "GET",
          "url": "http://localhost:<PORT>/hello",
          "httpVersion": "HTTP/1.1",
          "cookies": [],
          "headers": [
            {
              "name": "accept",
              "value": "*/*"
            },
            {
              "name": "host",
              "value": "localhost:<PORT>"
            },
            {
              "name": "content-length",
              "value": "0"
            }
          ],
          "queryString": [],
          "postData": {
            "mimeType": "",
            "text": ""
          },
          "headersSize": 86,
          "bodySize": 0
        },
        "response": {
          "status": 200,
          "statusText": "",
          "httpVersion": "HTTP/1.1",
          "cookies": [],
          "headers": [
            {
              "name": "content-length",
              "value": "13"
            },
            {
              "name": "date",
              "value": "<DATETIME>"
            }
          ],
          "content": {
            "size": 13,
            "mimeType": "",
            "text": "Hello, World!",
            "compression": 0
          },
          "redirectURL": "",
          "headersSize": 80,
          "bodySize": 13
        },
        "cache": {},
        "timings": {
          "connect": -1,
          "ssl": -1,
          "send": -1,
          "receive": -1,
          "wait": -1
        }
      },
      {
        "startedDateTime": "<DATETIME>",
        "time": <TIME>,
        "request": {
          "method": "GET",
          "url": "http://localhost:<PORT>/hello/gzip",
          "httpVersion": "HTTP/1.1",
          "cookies": [],
          "headers": [
            {
              "name": "accept",
              "value": "*/*"
            },
            {
              "name": "host",
              "value": "localhost:<PORT>"
            },
            {
              "name": "content-length",
              "value": "0"
            }
          ],
          "queryString": [],
          "postData": {
            "mimeType": "",
            "text": ""
          },
          "headersSize": 86,
          "bodySize": 0
        },
        "response": {
          "status": 200,
          "statusText": "",
          "httpVersion": "HTTP/1.1",
          "cookies": [],
          "headers": [
            {
              "name": "content-encoding",
              "value": "gzip"
            },
            {
              "name": "transfer-encoding",
              "value": "chunked"
            },
            {
              "name": "date",
              "value": "<DATETIME>"
            }
          ],
          "content": {
            "size": 33,
            "mimeType": "",
            "text": "Hello, World!",
            "compression": -20
          },
          "redirectURL": "",
          "headersSize": 120,
          "bodySize": 33
        },
        "cache": {},
        "timings": {
          "connect": -1,
          "ssl": -1,
          "send": -1,
          "receive": -1,
          "wait": -1
        }
      },
      {
        "startedDateTime": "<DATETIME>",
        "time": <TIME>,
        "request": {
          "method": "POST",
          "url": "http://localhost:<PORT>/echo",
          "httpVersion": "HTTP/1.1",
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "text/plain"
            },
            {
              "name": "accept",
              "value": "*/*"
            },
            {
              "name": "host",
              "value": "localhost:<PORT>"
            },
            {
              "name": "content-length",
              "value": "13"
            }
          ],
          "queryString": [],
          "postData": {
            "mimeType": "text/plain",
            "text": "Hello, World!"
          },
          "headersSize": 121,
          "bodySize": 13
        },
        "response": {
          "status": 200,
          "statusText": "",
          "httpVersion": "HTTP/1.1",
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "text/plain"
            },
            {
              "name": "content-length",
              "value": "13"
            },
            {
              "name": "date",
              "value": "<DATETIME>"
            }
          ],
          "content": {
            "size": 13,
            "mimeType": "text/plain",
            "text": "Hello, World!",
            "compression": 0
          },
          "redirectURL": "",
          "headersSize": 114,
          "bodySize": 13
        },
        "cache": {},
        "timings": {
          "connect": -1,
          "ssl": -1,
          "send": -1,
          "receive": -1,
          "wait": -1
        }
      }
    ]
  }
}

/traffics?curl

curl http://localhost:<PORT>/hello \
  -H 'accept: */*' \
  -H 'host: localhost:<PORT>'

curl http://localhost:<PORT>/hello/gzip \
  -H 'accept: */*' \
  -H 'host: localhost:<PORT>'

curl http://localhost:<PORT>/echo \
  -X POST \
  -H 'content-type: text/plain' \
  -H 'accept: */*' \
  -H 'host: localhost:<PORT>' \
  -d 'Hello, World!'
